grin/test/Transformations/SyntaxSpec.hs

```diff
@@ -0,0 +1,38 @@
+module Transformations.SyntaxSpec where
+
+import Control.DeepSeq
+
+import Grin.Grin
+import Grin.Syntax (Exp)
+import qualified Grin.Syntax.Extended as New (Exp)
+import Transformations.Syntax
+
+import Test.Hspec
+import Test.QuickCheck
+import Test.Hspec.QuickCheck
+
+import Test.Assertions
+import Test.ExtendedSyntax.Test()
+import qualified Test.ExtendedSyntax.Grammar as G
+
+runTests :: IO ()
+runTests = hspec spec
+
+spec :: Spec
+spec = describe "Syntax transformation QuickCheck tests" $ do
+         prop "Old is always convertible to New" $
+           convertibleToNew
+         prop "Old is always convertible to New then back to Old" $
+           roundtripConvertibleOld
+
+-- NOTE: The conversion itself is the proof that it is convertible
+-- QUESTION: There must be a better way to do this
+-- ANSWER: The conversion function could an Either
+convertibleToNew :: G.Exp -> Bool
+convertibleToNew exp = force (convertToNew $ G.asExp exp) `seq` True
+
+roundtripConvertibleOld :: G.Exp -> Bool
+roundtripConvertibleOld exp = force (convertToOld $ convertToNew $ G.asExp exp) `seq` True where
```

Comment:
Please use the `==` in this function, as the round-trip property gives another expression to compare against.

Reply:
You might not get back the same AST as you started with. `convertToNew` introduces new names, and new bindings. This property only tests whether the generated old AST does not contain any constructs that are not convertible to the new one, and that the converted new AST does not contain any constructs that are not convertible back to the old one.

Reply:
I am surprised, that the `hspec` does not have a `shouldNotThrow` check. As the `convertToNew` will be removed throwing an exception feels okish.